### PR TITLE
add distributed mode start

### PR DESCRIPTION
**Pull request by voyager-dw**
add separate cmd start path for distributed mode

### CODE REVIEW

**Comment by voyager-dw** [comment]
need more tests on bootstrap and datanode, we can add in next 2 weeks

File: cmd/aresd/cmd/cmd.go [hunk 1]
```diff
@@ -234,25 +243,47 @@ func start(cfg common.AresServerConfig, logger common.Logger, queryLogger common
...
+	if err != nil {
+		logger.Fatal("Failed to create datanode,", err)
+	}
+
+	err = dataNode.Open()
+	if err != nil {
+		logger.Fatal("Failed to open datanode,", err)
+	}
+	dataNode.Bootstrap()
+	if err != nil {
```
Comment (jshencode) [inline cmd/aresd/cmd/cmd.go]: err := dataNode.Bootstrap()?

File: datanode/datanode.go [hunk 2]
```diff
@@ -335,16 +377,24 @@ func (d *dataNode) Serve() {
...
+	grpcServer := grpc.NewServer()
+	rpc.RegisterPeerDataNodeServer(grpcServer, d.bootstrapServer)
+	reflection.Register(grpcServer)
+
 	// record time from data node started to actually serving
 	d.metrics.restartTimer.Record(utils.Now().Sub(d.startedAt))
 	d.opts.InstrumentOptions().Logger().Infof("Starting HTTP server on port %d with max connection %d", d.opts.ServerConfig().Port, d.opts.ServerConfig().HTTP.MaxConnections)
-	utils.LimitServe(d.opts.ServerConfig().Port, handlers.CORS(allowOrigins, allowHeaders, allowMethods)(router), d.opts.ServerConfig().HTTP)
+	utils.LimitServe(d.opts.ServerConfig().Port, handlers.CORS(allowOrigins, allowHeaders, allowMethods)(mixedHandler(grpcServer, router)), d.opts.ServerConfig().HTTP)
+	grpcServer.Stop()
```
Comment (jshencode) [inline datanode/datanode.go]: move these to Close()?

File: cmd/aresd/cmd/cmd.go [hunk 3]
```diff
@@ -15,14 +15,21 @@
 package cmd
 
 import (
+	"code.uber.internal/rt/onix/clients/logger"
 	"fmt"
 	"github.com/gorilla/handlers"
 	"github.com/gorilla/mux"
+	"github.com/m3db/m3/src/cluster/services"
+	"github.com/m3db/m3x/instrument"
```
Comment (jshencode) [inline cmd/aresd/cmd/cmd.go]: use m3/src/x/instrument instead

File: cmd/aresd/cmd/cmd.go (hunk 1 above)
Comment (voyager-dw) [reply cmd/aresd/cmd/cmd.go]: good catch

File: datanode/datanode.go (hunk 2 above)
Comment (voyager-dw) [reply datanode/datanode.go]: good idea

File: cmd/aresd/cmd/cmd.go [hunk 4]
```diff
@@ -234,25 +242,51 @@ func start(cfg common.AresServerConfig, logger common.Logger, queryLogger common
...
+// start datanode in distributed mode
+func startDataNode(cfg common.AresServerConfig, logger common.Logger, scope tally.Scope, httpWrappers ...utils.HTTPHandlerWrapper) {
+	serverRestartTimer := scope.Timer("restart").Start()
+
+	opts := datanode.NewOptions().SetServerConfig(cfg).SetInstrumentOptions(utils.NewOptions()).SetBootstrapOptions(bootstrap.NewOptions()).SetHTTPWrappers(httpWrappers)
+
+	var topo topology.Topology
+	etcdCfg := cfg.InstanceConfig.Etcd
+	etcdCfg.Service = utils.DataNodeServiceName(cfg.InstanceConfig.Namespace)
+	configServiceCli, err := etcdCfg.NewClient(instrument.NewOptions())
```
[the remark this answers is not in the document]
Comment (voyager-dw) [reply cmd/aresd/cmd/cmd.go]: will change later, they are a different option, need conversion

**Pull request merged**